src/plot_api/plot_api.js

```diff
@@ -2208,6 +2195,356 @@ Plotly.update = function update(gd, traceUpdate, layoutUpdate, _traces) {
     });
 };
 
+/**
+ * Plotly.react:
+ * A plot/update method that takes the full plot state (same API as plot/newPlot)
+ * and diffs to determine the minimal update pathway
+ *
+ * @param {string id or DOM element} gd
+ *      the id or DOM element of the graph container div
+ * @param {array of objects} data
+ *      array of traces, containing the data and display information for each trace
+ * @param {object} layout
+ *      object describing the overall display of the plot,
+ *      all the stuff that doesn't pertain to any individual trace
+ * @param {object} config
+ *      configuration options (see ./plot_config.js for more info)
+ *
+ * OR
+ *
+ * @param {string id or DOM element} gd
+ *      the id or DOM element of the graph container div
+ * @param {object} figure
+ *      object containing `data`, `layout`, `config`, and `frames` members
+ *
+ */
+Plotly.react = function(gd, data, layout, config) {
+    var frames, plotDone;
+
+    function addFrames() { return Plotly.addFrames(gd, frames); }
+
+    gd = Lib.getGraphDiv(gd);
+
+    var oldFullData = gd._fullData;
+    var oldFullLayout = gd._fullLayout;
+
+    // you can use this as the initial draw as well as to update
+    if(!Lib.isPlotDiv(gd) || !oldFullData || !oldFullLayout) {
+        plotDone = Plotly.newPlot(gd, data, layout, config);
+    }
+    else {
+
+        if(Lib.isPlainObject(data)) {
+            var obj = data;
+            data = obj.data;
+            layout = obj.layout;
+            config = obj.config;
+            frames = obj.frames;
+        }
+
+        var configChanged = false;
+        // assume that if there's a config at all, we're reacting to it too,
+        // and completely replace the previous config
+        if(config) {
+            var oldConfig = Lib.extendDeep({}, gd._context);
+            gd._context = undefined;
+            setPlotContext(gd, config);
+            configChanged = diffConfig(oldConfig, gd._context);
+        }
+
+        gd.data = data || [];
+        helpers.cleanData(gd.data, []);
+        gd.layout = layout || {};
+        helpers.cleanLayout(gd.layout);
+
+        Plots.supplyDefaults(gd);
+
+        var newFullData = gd._fullData;
+        var newFullLayout = gd._fullLayout;
+        var immutable = newFullLayout.datarevision === undefined;
+
+        var restyleFlags = diffData(gd, oldFullData, newFullData, immutable);
+        var relayoutFlags = diffLayout(gd, oldFullLayout, newFullLayout, immutable);
+
+        // clear calcdata if required
+        if(restyleFlags.calc || relayoutFlags.calc) gd.calcdata = undefined;
+
+        // Note: what restyle/relayout use impliedEdits and clearAxisTypes for
+        // must be handled by the user when using Plotly.react.
```

Comment:
⚠️ not sure if there's something we can do to make this easier for users (pull out the relevant sections from `_restyle` and `_relayout` perhaps into helper functions?), but I don't think we should be trying to do anything with `impliedEdits` or `clearAxisTypes` inside `Plotly.react`.

Reply:
Can you clarify what you mean by: _must be handling by user_?

Reply:
`impliedEdits` are things like turning off `axis.autorange` when you set an explicit `range`. If you use `relayout` et al we do that for you, so `Plotly.relayout(gd, 'yaxis.range', [5, 10])` automatically sets `yaxis.autorange: false` - otherwise the range wouldn't actually change, as autorange would put it back to what you started with.

The equivalent with `Plotly.react` would be:
```js
layout.yaxis.range = [5, 10];
// you have to explicitly disable autorange when using `Plotly.react`
// even if you never explicitly enabled it in the first place, since plotly.js
// inserted `yaxis.autorange: true` in your layout object
layout.yaxis.autorange = false;
Plotly.react(gd, data, layout);
```

`clearAxisTypes` is similar: using `Plotly.restyle(gd, 'x', [['a', 'b', 'c']], [0])` will clear the axis types of the axes trace 0 is displayed on, so the autotype algorithm runs again. With `Plotly.react` you'll have to explicitly remove `xaxis.type` if you want to invoke autotype.

I suppose there's a discussion to be had about this: it would certainly be nice to just make the change you want, the same as you do with `relayout`, and have us figure out what we need to do to make it happen. On the other hand, with `Plotly.react` you're saying "this is the new plot state I want rendered", and for us to comb through it and say "Oh, I see you made this change, you must have wanted this other change too" seems heavy-handed and error-prone - what if you really don't want that other change, there's no way to tell us that (as there would be with `relayout` eg `Plotly.relayout(gd, {'yaxis.range': [5, 10], 'yaxis.autorange': true})` (which seems a contrived example, but you might actually do that for example to turn the range from decreasing autorange to increasing autorange).

So rather than bake this into `Plotly.react` as we do with `relayout` et al, I think we should try and find a way to help the user access the recommendations we would make with `impliedEdits` and `clearAxisTypes`, but leave it up to them whether to follow those recommendations.

Reply:
> what if you really don't want that other change, there's no way to tell us that 

Thanks for writing this. This makes :100: % total sense :books: 

>  think we should try and find a way to help the user access the recommendations we would make with impliedEdits and clearAxisTypes, but leave it up to them whether to follow those recommendations.

Perhaps all we need is to (under a certain config option) log whenever we mutate user data or user layout. Alternatively, `Plotly.validate` could be use here e.g.

```js
Plotly.validate([{
  y: [1, 2, 1]
}], {
  xaxis: {
     autorange: true,
     range: [1, 2]
  }
})
```

could log something like: _increasing range detected, range values ignored under autorange_. Then together with https://github.com/plotly/plotly.js/issues/1741, it should make it fairly easy for plotly.js users to debug.